### PR TITLE
Issue/6850 add google signup

### PR DESCRIPTION
**Pull request by theck13**
### Fix
Add Google signup when the ***Sign Up With Google*** button is pressed on the prologue screen as described in https://github.com/wordpress-mobile/WordPress-Android/issues/6850.

These changes also extract the code specific to Google login and signup from the `LoginGoogleFragment` into a new `GoogleFragment` file to avoid code duplication.  Now, the code in the `LoginGoogleFragment` and `SignupGoogleFragment` files only relate to the Google login and signup flow, respectively.

### Test
***New Account***

0. Clear app data.
1. Tap ***Sign Up*** button.
2. Tap ***Sign Up with Google*** button.
3. Choose Google account without existing WordPress.com account.
4. Notice nothing happens.  The next step will be handling in a subsequent pull request.

***Old Account / Social Disconnected / Two-Factor Disabled***

0. Clear app data.
1. Tap ***Sign Up*** button.
2. Tap ***Sign Up with Google*** button.
3. Choose Google account with existing WordPress.com account.
4. Notice message explaining an account exists and continue with login.
5. Enter WordPress.com account password.
6. Tap ***Next*** button.
7. Notice ***Epilogue*** screen is shown.

***Old Account / Social Connected / Two-Factor Disabled***

0. Clear app data.
1. Tap ***Sign Up*** button.
2. Tap ***Sign Up with Google*** button.
3. Choose Google account with existing WordPress.com account.
4. Notice ***Epilogue*** screen is shown.

***Old Account / Social Disconnected / Two-Factor Enabled***

0. Clear app data.
1. Tap ***Sign Up*** button.
2. Tap ***Sign Up with Google*** button.
3. Choose Google account with existing WordPress.com account.
4. Notice message explaining an account exists and continue with login.
5. Enter WordPress.com account password.
6. Tap ***Next*** button.
7. Notice ***Two-Factor Authentication*** screen is shown.
8. Enter two-factor authentication code.
9. Tap ***Next*** button.
10. Notice ***Epilogue*** screen is shown.

***Old Account / Social Connected / Two-Factor Enabled***

0. Clear app data.
1. Tap ***Sign Up*** button.
2. Tap ***Sign Up with Google*** button.
3. Choose Google account with existing WordPress.com account.
4. Notice message explaining an account exists and continue with login.
5. Notice ***Two-Factor Authentication*** screen is shown.
6. Enter two-factor authentication code.
7. Tap ***Next*** button.
8. Notice ***Epilogue*** screen is shown.

#### Note
There are two copy changes.  One is the new timeout error message "Google took too long to respond. You may need to wait until you have a better connection." which occurs when Google login or signup fails due to taking too long.  The second is the addition of a message that appears when users are taken from the signup flow to the login flow.  It is an ephemeral message based off an [existing Google login error](https://github.com/wordpress-mobile/WordPress-Android/blob/issue/6850-add-google-signup/WordPress/src/main/res/values/strings.xml#L2008).  See the screenshot below for illustration.

### CODE REVIEW

**Comment by nbradbury** [comment]
This isn't working for me.  After I tap "Sign up with Google," it takes me through a flow that adds the account to the **device** rather than the app. Repro'd with two emulators - API 22 and API 26.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginGoogleFragment.java [hunk 1]
```diff
@@ -1,299 +1,124 @@
...
                             // Attempt didn't succeed with the current account.
                             case GoogleSignInStatusCodes.SIGN_IN_FAILED:
-                                AppLog.e(T.NUX, "Google Sign-in Failed: current account failed.");
+                                AppLog.e(T.NUX, "Google Login Failed: current account failed.");
                                 showErrorDialog(getString(R.string.login_error_generic));
                                 break;
                             // Attempted to connect, but the user is not signed in.
                             case GoogleSignInStatusCodes.SIGN_IN_REQUIRED:
-                                AppLog.e(T.NUX, "Google Sign-in Failed: user is not signed in.");
+                                AppLog.e(T.NUX, "Google Login Failed: user is not signed in.");
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginGoogleFragment.java]: [The docs](https://developers.google.com/android/reference/com/google/android/gms/common/ConnectionResult.html#SIGN_IN_REQUIRED) are confusing, but if I'm reading correctly then when `SIGN_IN_REQUIRED` is returned we may be able to continue with sign in (?)

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginGoogleFragment.java [hunk 2]
```diff
@@ -1,299 +1,124 @@
...
                                 showErrorDialog(getString(R.string.login_error_generic));
                                 break;
                             // Unknown error.
                             default:
-                                AppLog.e(T.NUX, "Google Sign-in Failed: unknown error.");
+                                AppLog.e(T.NUX, "Google Login Failed: unknown error.");
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginGoogleFragment.java]: Should we handle `GoogleSignInStatusCodes.TIMEOUT` separately?

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupGoogleFragment.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                                break;
+                            // Unknown error.
+                            default:
+                                AppLog.e(T.NUX, "Google Signup Failed: unknown error.");
+                                showErrorDialog(getString(R.string.login_error_generic));
+                                break;
+                        }
+                    }
+                } else if (result == RESULT_CANCELED) {
+                    AnalyticsTracker.track(AnalyticsTracker.Stat.LOGIN_SOCIAL_BUTTON_FAILURE);
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupGoogleFragment.java]: We're tracking `LOGIN_SOCIAL_BUTTON_FAILURE` with no extra data both here and below, so we won't be able to tell if the user cancelled or something else went wrong. Should we be doing that?

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupGoogleFragment.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    public void onSocialChanged(AccountStore.OnSocialChanged event) {
+        if (event.isError()) {
+            AppLog.e(T.API, "SignupGoogleFragment.onSocialChanged: " + event.error.type + " - " + event.error.message);
+
+            switch (event.error.type) {
+                // WordPress account exists with input email address, and two-factor authentication is required.
+                case TWO_STEP_ENABLED:
+                    AnalyticsTracker.track(AnalyticsTracker.Stat.SIGNUP_SOCIAL_2FA_NEEDED);
+                    ToastUtils.showToast(getContext(), getString(R.string.signup_user_exists, mGoogleEmail),
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupGoogleFragment.java]: Can we shorten the length of `signup_user_exists`? Even with a long toast, it's not easy to read the whole thing before it disappears, especially when you're not expecting it.

Also, the phrase "continue with login" is a bit cryptic and/or unfriendly. Might want to ask Editorial for better copy.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginGoogleFragment.java (hunk 1 above)
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginGoogleFragment.java]: Yeah, the terminology is a little confusing.  We've tried to differentiate by using _login_ when referring to an existing account and _signup_ when talking about creating a new account.  The term sign-in used by the API and Google documentation refers to both login and signup depending on the situation, but the `SIGN_IN_REQUIRED` in this case means login.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginGoogleFragment.java (hunk 2 above)
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginGoogleFragment.java]: We didn't discuss it in the initial designs, but it seems like a good idea.  I added that case in for login and signup 12ea33fe1b300f911725eaf6c57780bc486e8e37.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupGoogleFragment.java (hunk 3 above)
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupGoogleFragment.java]: Unfortunately, `RESULT_CANCELED` can be multiple things (e.g. canceled by the user intentionally or canceled by the Google API due to some unknown reason).  We don't get anything other than the result value so we can't really handle it any differently than an unknown error like in the `else` case below.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupGoogleFragment.java (hunk 4 above)
Comment (theck13) [reply WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SignupGoogleFragment.java]: Yep, I totally agree.  I meant to add the `[Status] Needs Copy Review` label for wordsmithing help from Editorial.  I'll add that and `[Status] Needs Design Review`.

**Comment by michelleweber** [comment]
This message feels ambiguous -- continue with the current process? Or go back to the login screen?

**Comment by theck13** [comment]
> This message feels ambiguous -- continue with the current process? Or go back to the login screen?

I'm guessing you're referring to the "The Google account 'heckofatest@gmail.com' matches an existing account on WordPress.com. Continue with login." message when the user is taken from the signup flow to the login flow, right?  The user is automatically taken to the login flow when this message is shown.  We wanted to add something to inform the user they are no longer in the signup flow to minimize confusion.  Do you know how we can express that better?

What do you think about the "Google took too long to respond. You may need to wait until you have a better connection." message shown when a timeout error occurs while trying to login or signup with Google?  See the screenshot below for illustration.

**Comment by nbradbury** [comment]
@theck13 I'm finished reviewing this and apart from the wording it all looks good. We can wait to merge until after the wording is finalized, or I can merge now and you could change the wording in the next PR - just let me know which you prefer.

**Comment by michelleweber** [comment]
The "Google took too long" message is good. I'd maybe say "stronger internet connection" instead of just "better connection."

For the other, I'm thinking something like "There's already a WordPress.com account using [email address] -- log in normally with your WordPress.com username and password."

Also -- on the login screen behind, it'd be great to change that second sentence to "We'll only ask for this once."

**Comment by theck13** [comment]
Thanks, @michelleweber!

> The "Google took too long" message is good. I'd maybe say "stronger internet connection" instead of just "better connection."

I'll update the timeout message to "Google took too long to respond. You may need to wait until you have a stronger internet connection."

> For the other, I'm thinking something like "There's already a WordPress.com account using [email address] -- log in normally with your WordPress.com username and password."

I can change the message to that, but this message is also shown when the user is taken to the two-factor authentication screen, which has only a verification code field rather than username and password fields.  Should we have one message for both cases or unique messages for each case?

**Comment by folletto** [comment]
Ok so if I understand correctly there are two messages to review:

1. When the connection to Google auth server times out. 
2. When the user trying to signup is moved instead to login.

Also:

1. Is a modal window with just close as action.
2. Can we make this a snackbar at the bottom?

Given the above, I'd suggest the following:

**Timeout—**

* Modal with "Back" button ("Close" is ok too if you prefer that)
* "Unable to connect to Google. Try again later."
* The reason for this copy is that while it's likely for being a connection error, a timeout could also happen on Google side, or due to routing issues of some kind, so I would suggest to make the user wait a bit instead of suggesting it's a fault of the connection.

**Signup tp Login—**

* Snackbar at the bottom, with no action. Disappears after the default amount of time (5 seconds?). Can we?
* "Email already exists on WordPress.com. Proceeding with log in."

The above still pending copy review of course. ;)

**Comment by michelleweber** [comment]
> I can change the message to that, but this message is also shown when the user is taken to the two-factor authentication screen, which has only a verification code field rather than username and password fields. Should we have one message for both cases or unique messages for each case?

Unique, I think. Any instruction/guidance we give should be as specific to what the user is currently seeing as possible, so I'd want the 2FA messaging to be about entering a 2FA code to continue login.

**Comment by theck13** [comment]
@folletto:
> Ok so if I understand correctly there are two messages to review:
> 
> 1. When the connection to Google auth server times out.
> 2. When the user trying to signup is moved instead to login.

Yes, those are the two scenarios and both of your suggestions are possible as you described them.

@michelleweber:
How is the wording @folletto described above?  One thing to note about the suggested design is the  length of the messages.  The timeout dialog is meant to contain a sentence or two summarizing the situation.  The signup-to-login snackbar should be a very short text string containing a single line of text directly related to the operation performed.

**Comment by theck13** [comment]
@folletto and @michelleweber:
When testing the "Email already exists on WordPress.com. Proceeding with login." on the Pixel and Pixel 2, the message is actually two lines.  Would "Email already exists. Proceeding with login." work or something similar in length to keep the message to a single line?

**Comment by folletto** [comment]
> the message is actually two lines

Yeh, I'd break on two lines on the ".": 

> Email already exists on WordPress.com. 
> Proceeding with log in.

I think given it's an automated message out of the part of the screen that the user is already looking at, having two lines raised a bit its visual profile. :)

**Comment by theck13** [comment]
I was suggesting keeping the message to a single line to follow the [design guidelines](https://material.io/guidelines/components/snackbars-toasts.html#snackbars-toasts-usage).  Just to be clear, you are suggesting forcing the message to be two lines intentionally?

**Comment by folletto** [comment]
Yes, the guidelines include two lines support (`80dp`):

components-toasts-mobile4

**Comment by theck13** [comment]
Hurray for contradictory guidelines!

> Snackbars should contain a single line of text directly related to the operation performed.

> Multi-line snackbar height: 80dp

I'll use the "Email already exists on WordPress.com. Proceeding with login." and force the line break after the first period so that it always looks like this.

snackbar

**Comment by folletto** [comment]
Hehe yeh. I think they updated in two steps and never revised the whole doc. :D 

> I'll use the "Email already exists on WordPress.com. Proceeding with login." and force the line break after the first period so that it always looks like this.

Cool!

**Comment by michelleweber** [comment]
These messages are good. My preference is always to go more natural-speech-y, but space doesn't always allow it!

**Comment by nbradbury** [comment]
Looks good, let's :shipit: !

**Pull request merged**